akka-typed/src/main/scala/akka/typed/persistence/scaladsl/PersistentActor.scala

```diff
@@ -32,125 +32,186 @@ object PersistentActor {
+      new CompositeEffect[Event, State](Some(new PersistAll[Event, State](events)), sideEffects)
+
+    /**
+     * Do not persist anything
+     */
+    def done[Event, State]: Effect[Event, State] = PersistNothing.asInstanceOf[Effect[Event, State]]
+
+    /**
+     * This command is not handled, but it is not an error that it isn't.
+     */
+    def unhandled[Event, State]: Effect[Event, State] = Unhandled.asInstanceOf[Effect[Event, State]]
+
+    /**
+     * Stop this persistent actor
+     */
+    def stop[Event, State]: ChainableEffect[Event, State] = Stop.asInstanceOf[ChainableEffect[Event, State]]
+
+  }
 
-  sealed abstract class Effect[+Event, State]() {
+  /**
+   * Instances are created through the factories in the [[Effect]] companion object.
+   *
+   * Not for user extension.
+   */
+  @DoNotInherit
+  sealed abstract class Effect[+Event, State] {
     /* All events that will be persisted in this effect */
     def events: im.Seq[Event] = Nil
 
     /* All side effects that will be performed in this effect */
-    def sideEffects: im.Seq[ChainableEffect[_, State]] =
-      if (isInstanceOf[ChainableEffect[_, State]]) im.Seq(asInstanceOf[ChainableEffect[_, State]])
-      else Nil
-
-    def andThen(sideEffects: ChainableEffect[_, State]*): Effect[Event, State] =
-      CompositeEffect(if (events.isEmpty) None else Some(this), sideEffects.toList)
+    def sideEffects[E >: Event]: im.Seq[ChainableEffect[E, State]] = Nil
 
     /** Convenience method to register a side effect with just a callback function */
     def andThen(callback: State ⇒ Unit): Effect[Event, State] =
-      andThen(SideEffect[Event, State](callback))
+      CompositeEffect(this, SideEffect[Event, State](callback))
 
     /** Convenience method to register a side effect with just a lazy expression */
     def andThen(callback: ⇒ Unit): Effect[Event, State] =
-      andThen(SideEffect[Event, State]((_: State) ⇒ callback))
+      CompositeEffect(this, SideEffect[Event, State]((_: State) ⇒ callback))
+
+    /** The side effect is to stop the actor */
+    def andThenStop[E >: Event]: Effect[Event, State] =
+      CompositeEffect(this, Effect.stop[Event, State])
   }
 
-  case class CompositeEffect[Event, State](persistingEffect: Option[Effect[Event, State]], override val sideEffects: im.Seq[ChainableEffect[_, State]]) extends Effect[Event, State] {
-    override val events = persistingEffect.map(_.events).getOrElse(Nil)
-    override def andThen(additionalSideEffects: ChainableEffect[_, State]*): CompositeEffect[Event, State] =
-      copy(sideEffects = sideEffects ++ additionalSideEffects)
+  @InternalApi
+  private[akka] object CompositeEffect {
+    def apply[Event, State](effect: Effect[Event, State], sideEffects: ChainableEffect[Event, State]): Effect[Event, State] =
+      CompositeEffect[Event, State](
+        if (effect.events.isEmpty) None else Some(effect),
+        sideEffects.asInstanceOf[im.Seq[ChainableEffect[Event, State]]])
```

Comment:
is this cast correct?
isn't it `sideEffects: ChainableEffect[Event, State]`?

Reply:
Yup, ups, good catch